In [1]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler

In [2]:
train_df = pd.read_csv('Resources/2019loans.csv')
test_df = pd.read_csv('Resources/2020Q1loans.csv')

In [3]:
print(len(train_df))
print(len(test_df))

12180
4702


In [4]:
set(train_df.columns) == set(test_df.columns)

True

In [5]:
# Convert categorical data to numeric and separate target feature for training data
X_train = pd.get_dummies(train_df.drop(columns = "target"))
X_train

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0         7000.0    0.1894       256.38     75000.0  28.62          0.0   
1        40000.0    0.1614       975.71    102000.0  11.72          2.0   
2        11000.0    0.2055       294.81     45000.0  37.25          1.0   
3         4000.0    0.1612       140.87     38000.0  42.89          1.0   
4        14000.0    0.1797       505.93     43000.0  22.16          1.0   
...          ...       ...          ...         ...    ...          ...   
12175    19975.0    0.2565       801.09     28000.0  28.42          0.0   
12176    15000.0    0.1774       540.34     50000.0  23.43          4.0   
12177     3600.0    0.1862       131.28     60000.0  28.80          0.0   
12178    15000.0    0.0881       475.68     62000.0  11.44          0.0   
12179    15000.0    0.1774       540.34     54080.0  21.66          1.0   

       inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0                 2.0      20.0      0.0    40414.0  ...   
1                 0.0      10.0      0.0    43531.0  ...   
2                 3.0      23.0      0.0     8242.0  ...   
3                 0.0       7.0      0.0    12767.0  ...   
4                 0.0      22.0      0.0    11182.0  ...   
...               ...       ...      ...        ...  ...   
12175             0.0      15.0      0.0     9679.0  ...   
12176             0.0      16.0      0.0     3193.0  ...   
12177             1.0      14.0      0.0    11804.0  ...   
12178             0.0       5.0      0.0      157.0  ...   
12179             1.0       5.0      0.0     8349.0  ...   

       verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                                 0             1                      1   
1                                 0             1                      0   
2                                 1             1                      0   
3                                 0             1                      0   
4                                 0             1                      0   
...                             ...           ...                    ...   
12175                             0             1                      0   
12176                             1             1                      0   
12177                             0             1                      0   
12178                             0             1                      0   
12179                             0             1                      1   

       initial_list_status_w  application_type_Individual  \
0                          0                            1   
1                          1                            1   
2                          1                            1   
3                          1                            0   
4                          1                            1   
...                      ...                          ...   
12175                      1                            1   
12176                      1                            1   
12177                      1                            1   
12178                      1                            0   
12179                      0                            1   

       application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                               0                1                0   
1                               0                1                0   
2                               0                1                0   
3                               1                1                0   
4                               0                1                0   
...                           ...              ...              ...   
12175                           0                1                0   
12176                           0                1                0   
12177                           0                1                0   
12178                        

In [6]:
y_train = pd.get_dummies(train_df["target"])["high_risk"]
y_train

0        0
1        0
2        0
3        0
4        0
        ..
12175    1
12176    1
12177    1
12178    1
12179    1
Name: high_risk, Length: 12180, dtype: uint8

In [7]:
# Convert categorical data to numeric and separate target feature for testing data
X_test = pd.get_dummies(test_df.drop(columns = "target"))
X_test

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0       40000.0    0.1033       856.40    128700.0  12.47          0.0   
1       24450.0    0.1430       572.72     44574.0  15.05          0.0   
2       13500.0    0.1430       316.23     60000.0  28.72          0.0   
3       10625.0    0.1774       268.31     60000.0  15.70          0.0   
4        6375.0    0.1862       232.46     60000.0  35.50          0.0   
...         ...       ...          ...         ...    ...          ...   
4697    30000.0    0.1240       673.42    140480.0  15.74          0.0   
4698    24000.0    0.0756       747.22     50000.0  26.81          0.0   
4699    10000.0    0.2305       387.36     33000.0  38.51          0.0   
4700     8000.0    0.1862       205.86     38000.0  16.36          0.0   
4701    30000.0    0.2055      1123.34    180000.0  12.06          0.0   

      inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0                1.0       8.0      0.0    38113.0  ...   
1                1.0       6.0      0.0     1665.0  ...   
2                0.0       8.0      0.0    13857.0  ...   
3                4.0      17.0      0.0     6216.0  ...   
4                0.0      13.0      0.0    12681.0  ...   
...              ...       ...      ...        ...  ...   
4697             0.0      20.0      0.0    23215.0  ...   
4698             0.0       9.0      0.0      459.0  ...   
4699             2.0       7.0      0.0     6342.0  ...   
4700             1.0       8.0      1.0    11636.0  ...   
4701             0.0       8.0      0.0     4771.0  ...   

      verification_status_Source Verified  verification_status_Verified  \
0                                       1                             0   
1                                       0                             0   
2                                       0                             0   
3                                       0                             1   
4                                       1                             0   
...                                   ...                           ...   
4697                                    1                             0   
4698                                    0                             0   
4699                                    0                             1   
4700                                    1                             0   
4701                                    1                             0   

      pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0                1                      0                      1   
1                1                      0                      1   
2                1                      0                      1   
3                1                      0                      1   
4                1                      0                      1   
...            ...                    ...                    ...   
4697             1                      1                      0   
4698             1                      0                      1   
4699             1                      1                      0   
4700             1                      0                      1   
4701             1                      1                      0   

      application_type_Individual  application_type_Joint App  \
0                               1                           0   
1                               1                           0   
2                               1                           0   
3                               1                           0   
4                               1                           0   
...                           ...                         ...   
4697                            1                           0   
4698                            1                           0   
4699                            1                           0   
4700                            1       

In [8]:
y_test = pd.get_dummies(test_df["target"])["high_risk"]
y_test

0       0
1       0
2       0
3       0
4       0
       ..
4697    1
4698    1
4699    1
4700    1
4701    1
Name: high_risk, Length: 4702, dtype: uint8

In [9]:
set(X_train.columns) == set(X_test.columns)

False

In [10]:
train_cols = X_train.columns
test_cols = X_test.columns

print(len(train_cols))
print(len(test_cols))

92
91


In [11]:
for x in train_cols:
    if x not in test_cols:
        print(x)

debt_settlement_flag_Y


In [12]:
for y in test_cols:
    if y not in train_cols:
        print(y)

In [13]:
# add missing dummy variables to testing set
X_test["debt_settlement_flag_Y"] = 0

In [14]:
set(X_train.columns) == set(X_test.columns)

True

Random forest should perform better on the unscaled data. The magnitude of values in columns such as "loan_amt" and "annual_inc" might be more likely to distort the importance of these columns in a logistic regression model, when compared with the amount of variation in the categorical columns that have been converted to contain only values of 0 and 1. Random forest should be able to handle the upper range of values more effectively.

In [15]:
# Train the Logistic Regression model on the unscaled data and print the model score
clf = LogisticRegression(max_iter = 5000).fit(X_train, y_train)
print(f"Training Score: {clf.score(X_train, y_train)}")
print(f"Testing Score: {clf.score(X_test, y_test)}")

Training Score: 0.7038587848932677
Testing Score: 0.5642279880901744


In [16]:
# Train a Random Forest Classifier model and print the model score
rf = RandomForestClassifier(random_state = 1).fit(X_train, y_train)
print(f"Training Score: {rf.score(X_train, y_train)}")
print(f"Testing Score: {rf.score(X_test, y_test)}")

Training Score: 1.0
Testing Score: 0.6384517226712038


In [17]:
# Scale the data
scaler = StandardScaler().fit(X_train)

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

Logistic regression should perform better on scaled data because the variation within each column will be treated on a more comparable level. Having such a large number of columns makes it more likely for random forest to overfit the training data, and scaling is unlikely to change the patterns that led to overfitting the unscaled data.

In [18]:
# Train the Logistic Regression model on the scaled data and print the model score
clf_s = LogisticRegression(max_iter = 5000).fit(X_train_scaled, y_train)
print(f"Training Score: {clf_s.score(X_train_scaled, y_train)}")
print(f"Testing Score: {clf_s.score(X_test_scaled, y_test)}")

Training Score: 0.710919540229885
Testing Score: 0.7598894087622289


In [19]:
# Train a Random Forest Classifier model on the scaled data and print the model score
rf_s = RandomForestClassifier(random_state = 1).fit(X_train_scaled, y_train)
print(f"Training Score: {rf_s.score(X_train_scaled, y_train)}")
print(f"Testing Score: {rf_s.score(X_test_scaled, y_test)}")

Training Score: 1.0
Testing Score: 0.6384517226712038


Scaling the data led to an improvement from 0.56 to 0.76 in the test score of the logistic regression model, but no improvement in the random forest model score of 0.64. Random forest achieved an accuracy score of 1.0 on the training data, with and without scaling. The test scores of the random forest models are much lower than the training scores, indicating that the model is overfitted. 